flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManager.java

```diff
@@ -0,0 +1,287 @@
+	}
+
+	@Override
+	protected void onWorkerRegistered(WorkerType worker) {
+		final ResourceID resourceId = worker.getResourceID();
+		log.info("Worker {} is registered.", resourceId);
+
+		final WorkerResourceSpec workerResourceSpec = currentAttemptUnregisteredWorkers.remove(resourceId);
+		if (workerResourceSpec != null) {
+			final int count = pendingWorkerCounter.decreaseAndGet(workerResourceSpec);
+			log.info("Worker {} with resource spec {} was requested in current attempt." +
+							" Current pending count after registering: {}.",
+					resourceId,
+					workerResourceSpec,
+					count);
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	//  ResourceEventListener
+	// ------------------------------------------------------------------------
+
+	@Override
+	public void onPreviousAttemptWorkersRecovered(Collection<WorkerType> recoveredWorkers) {
+		log.info("Recovered {} workers from previous attempt.", recoveredWorkers.size());
+		for (WorkerType worker : recoveredWorkers) {
+			final ResourceID resourceId = worker.getResourceID();
+			workerNodeMap.put(resourceId, worker);
+			log.info("Worker {} recovered from previous attempt.", resourceId);
+		}
+	}
+
+	@Override
+	public void onWorkerTerminated(ResourceID resourceId) {
+		log.info("Worker {} is terminated.", resourceId);
+		if (clearStateForWorker(resourceId)) {
+			requestWorkerIfRequired();
+		}
+	}
+
+	@Override
+	public void onError(Throwable exception) {
+		onFatalError(exception);
+	}
+
+	@Override
+	public void handleInMainThread(Runnable runnable) {
+		runAsync(runnable);
+	}
+
+	// ------------------------------------------------------------------------
+	//  Internal
+	// ------------------------------------------------------------------------
+
+	private void requestNewWorker(WorkerResourceSpec workerResourceSpec) {
+		final TaskExecutorProcessSpec taskExecutorProcessSpec =
+				TaskExecutorProcessUtils.processSpecFromWorkerResourceSpec(flinkConfig, workerResourceSpec);
+		final int pendingCount = pendingWorkerCounter.increaseAndGet(workerResourceSpec);
+
+		log.info("Requesting new worker with resource spec {}, current pending count: {}.",
+				workerResourceSpec,
+				pendingCount);
+
+		resourceManagerDriver.requestResource(taskExecutorProcessSpec)
+				.whenComplete((worker, exception) -> {
+					if (exception != null) {
+						final int count = pendingWorkerCounter.decreaseAndGet(workerResourceSpec);
+						log.warn("Failed requesting worker with resource spec {}, current pending count: {}, exception: {}",
+								workerResourceSpec,
+								count,
+								exception);
+						requestWorkerIfRequired();
+					} else {
+						final ResourceID resourceId = worker.getResourceID();
+						workerNodeMap.put(resourceId, worker);
+						currentAttemptUnregisteredWorkers.put(resourceId, workerResourceSpec);
+						log.info("Requested worker {} with resource spec {}.",
+								resourceId,
+								workerResourceSpec);
+					}
+				});
```

Comment:
It would be good to guard against unwanted exceptions via `FutureUtils.assertNoException(...)`. In this case `whenComplete` would have to become `handle` which returns `null`.

Reply:
I'm not sure whether I have understand your comment correctly.
It seems `FutureUtils.assertNoException` will treat the exception as a fatal error and exit the process, while here what we want is to try re-request a worker if there's an exception.

Reply:
`FutureUtils.assertNoException` will guard against any unforeseen exceptions which might bubble up from the `requestWorkerIfRequired` method, for example. The problem with the current version is that these exception will get silently swallowed by the resulting future of `whenComplete`.

In other words, I don't want to treat the exception coming out of `requestResource` as fatally but all other exceptions coming out of the `whenComplete` handler because they won't be handled properly.

Reply:
Ah, I see. Thanks for the explanation.